ODIN_II/SRC/generate_blif_stats.cpp

```diff
@@ -0,0 +1,1128 @@
+static unsigned int pad_fanout_live = 0;
+static unsigned int vcc_fanout_all = 0;
+static unsigned int gnd_fanout_all = 0;
+static unsigned int pad_fanout_all = 0;
+static unsigned int max_input_pathrange = 0; // this is like reverse net_range bc it is dist to output not depth, abs(min-max pathlength)
+static unsigned int min_input_pathrange = 0-1;
+static float avg_pathrange = 0;
+static unsigned int num_nodes = 0;
+static unsigned int num_carry_func = 0;
+static unsigned int num_names = 0; // consider renaming this to num_gates bc that's what it is
+static unsigned int num_nets = 0; // same as output pin total it seems
+static unsigned int num_edges_fo = 0; // fan out total for all nets (includes vcc and gnd, unclear if it should, abc does not)
+static unsigned int num_edges_fi = 0; // fan in total, will equal num_total_input_pins and thus is redundant
+static unsigned int num_logic_nodes = 0;
+static unsigned int num_total_input_pins = 0; // reflects number of edges in graph that are live, should equal num_edges_fo but may not if there are fo pins w/out node 
+static unsigned int output_pin_total = 0; // this (should) always be the same as num_nets, maybe redundant
+static unsigned int num_subckts = 0;
+static unsigned int num_muxes = 0;
+static unsigned int num_generic_nodes = 0;
+static unsigned int num_input_nodes = 0;
+static unsigned int top_input_fanout = 0;
+static unsigned int num_unconn = 0; // Unreliable, maybe count pad traversals
+static unsigned int num_clock_nodes = 0;
+static unsigned int num_adders = 0;
+
+// This is all metadata on the number of traversals it took to calculate these figures, not passed to parser
+static unsigned int num_traversals_stat = 0; // how many depth first traversals have we done? Internal stat metric to measure cost of collection
+static unsigned int num_traversals_input = 0; // only counts traversals from top inputs (derived value not counter)
+static unsigned int gnd_vcc_pad_traverals = 0; // metadata on measuring gnd and vcc traversals, should be minimal if non-verbose and after all other traversals
+static unsigned int num_clearing_traversals = 0; // node traversals from clearing stat structs
+
+//static unsigned int num_unique_paths = 0; // Experimental, number of unique paths, this would be hard to calculate
+static unsigned int fanin_outputs_total = 0; // this is possibly not a useful stat
+//static unsigned int longest_chain = 1; // we're looking at latches but one might want adders or carries etc.
+//static unsigned int longest_consecutive_chain = 1;
+static unsigned int fanin_node_to_output = 0; // not very interesting
+static unsigned int fanout_node_to_output = 0; // experimental
+static unsigned int num_nodes_leading_to_output = 0; // unless we have fanout in net this will equal fanin_outputs_total, isn't v interesting but needed to derive
+static unsigned int num_deadends = 0; // experimental, every time we hit a deadEnd in path not including already visited nodes
```

Comment:
these are actually very bad to have since anything that does not drive an output node should be cleaned.   we should warn to do a cleanup phase beforehand

Reply:
Deadends only seem to be encountered when using a .v file as input instead of a BLIF, these deadends then seem to get cleaned up as part of the process of writing out to a BLIF (and thus is not reflected in this stat profiling). I'm not sure if it's in the spirit of profiling a netlist to then alter it however, although I can have it throw warnings on encountering a deadend.

Also it counts top input nodes (excluding pad) with no outgoing edges as a form of deadend and thus always considers the global_sim_base_clock as a deadend, I'm not sure if this should be counted as such.